### ISSUE TITLE
LSP: need a strategy for handling stateful data

### ISSUE THREAD

**Issue by ofseed** [neovim/neovim#31453]
### Problem

Some LSP request responses need to be retained after processing the results, such as:

- `textDocument/foldingRange` needs to retain state for closing a specific `foldingRangeKind`. For neovim, folding information must be cached.
- `textDocument/inlayHint` and `textDocument/codeLens` require maintaining state for subsequent requests like `resolve`.
- `textDocument/documentHighlight` could retain state to enable forward/backward navigation.

### Expected behavior

Implement an internal-only module dedicated to handling these stateful data, by:

- Creating `bufstate` and corresponding callback functions when a buffer has a client that supports the relevant methods.
- Destroying `bufstate` and corresponding callback functions when no clients in the buffer support the relevant methods.

Use this module for the methods mentioned above to avoid relying on nil-checks (which might mask bugs) and automatically creating tables via metatable methods (which can be difficult to debug).

The `bufstate` should only contain results directly obtained from the server (e.g., `lsp.foldingRange[]`, `lsp.InlayHint[]`) and derived information (e.g.,  computed caches or whether they are applied). It should not store user-managed settings (e.g., `enabled`).

**Comment by ofseed** [comment]
If this is considered a viable strategy, I would like to convert this issue into a tracking issue and proceed with refactoring related functionalities, starting with `lsp/_folding_range.lua`, then `lsp/codelens.lua`, finally `lsp/inlay_hint.lua`.

**Comment by justinmk** [comment]
Will be helpful to to existing code that will be replaced/lifted into this architecture.

> Use this module for the methods mentioned above to avoid relying on nil-checks (which might mask bugs) and automatically creating tables via metatable methods (which can be difficult to debug).

Sounds like a benefit. What are the costs, i.e. why wouldn't we do this?

p.s. is https://github.com/neovim/neovim/issues/31454 related?

**Comment by ofseed** [comment]
> What are the costs, i.e. why wouldn't we do this?

My concern is the risk of regression of already stable and usable features, as refactoring involves some rewriting of the code. My familiarity with the code is not as high as that of the maintainers, so I am unsure if I have enough experience to address this issue, but I am willing to try it myself if others do not. Therefore, the roadmap I currently have is relatively low-risk:

1. Refine the state management part in `_folding_range.lua` and extract state management into a separate module/function.
2. Change the API of `codelens.lua` to be similar to `inlay_hint.lua`, that is, `vim.lsp.codelens.enable()`.

I was already very interested in the goals above before my recent PRs, but it seems that they are currently blocked by the state management strategy. The priority for refactoring `inlay_hint.lua` is not high, as it should currently be a relatively stable API, and there is no compelling reason to refactor it at this time.

> p.s. is https://github.com/neovim/neovim/issues/31454 related?

I hadn't looked at `vim.lsp.diagnostic` in detail before, but it seems to be quite relevant. Interestingly, I mentioned in the chat that I experienced significant lag when modifying `vim.lsp.buf`, which seems to be the reason described in this issue. I can explore further, and if it is indeed related, we can solve it together.

**Comment by ribru17** [comment]
Would it be bad to keep folding data in a more generalized place (such that it is not specific to LSP folds)? I dream of a way in the future to set something like 

```scm
((import_statement)+ @fold 
  (#set! foldkind.imports))
```

using tree-sitter queries, and then users can just set some simple config option to auto-fold imports which will work regardless of whether a certain filetype is using TS or LSP folding

I assume this is relevant because the current LSP way does this by iterating through the LSP's own cache of folds, checking fold type (and TS has its own separate one fold cache)

**Comment by lewis6991** [comment]
I think mixing folding for both LSP and treesitter would be way too complex, and thus out of scope, unless there is a strong argument. These are just different backends

**Comment by ofseed** [comment]
> Would it be bad to keep folding data in a more generalized place (such that it is not specific to LSP folds)? I dream of a way in the future to set something like

My personal recommendation is to implement a server with a backend based on tree-sitter, providing folding data through LSP (after all, it is an interface). I should have implemented basic support for multiple servers. This may not be closely related to the scope of this issue. I mainly want to find a reliable way to save the result of the LSP response (not only folding ranges but also inlay hints or code lenses) instead of handling it and immediately discarding it here.

**Comment by ofseed** [comment]
Among all the implemented features that require state management (including semantic highlighting, diagnostics, inlay hints, and folding range), the model we use is that buffer data has client data to facilitate garbage collection when the buffer is deleted.

However, https://github.com/neovim/neovim/issues/32247 indicates that there may be some client data in LSP that needs to persist even after the buffer is deleted. This makes me a bit worried about thoroughly addressing this issue based on the current model.

Fortunately, the issue reported in the above issue seems to be a flaw of LSP itself. I suspect it may be due to the fact that vscode's OOP model manages data primarily through the client class, which may lead them to adopt a strategy where client data owns buffer data. This ambiguous and flawed definition in LSP is very likely something they added as a private addition to directly describe vscode's erroneous behavior.

Leave this comment as a reminder.

**Comment by justinmk** [comment]
Copying comment from https://github.com/neovim/neovim/discussions/34254#discussioncomment-13356500 : 

> introduce an internal module (or modify `lsp/client.lua`) that enables each LSP client to store independent buffer-level data. The key challenge is ensuring proper cleanup when a client detaches or a buffer is deleted, to prevent memory leaks.

The main goal is to find *something* that can be shared amongst all the existing use-cases in the lsp modules. Initially, it will just be a frankenstein in `runtime/lua/vim/lsp/util.lua` or `_util.lua` , and internal-only. This initial step will tell us whether it can be further generalized. Then we can refine the interface.

Going a bit further, it's worth seeing if https://github.com/neovim/neovim/issues/12544#issuecomment-1116794687 could be useful here, and maybe we want to implement that in this gsoc project.

We want:

- the "key" for stored items is caller-defined. some callers may want to key on namespace+bufnr, others may need something else?
- cleanup on buf-delete and other events (as you mentioned)
- no memory leaks; Lua "weak table" may be useful.
- visibility (so :checkhealth can print something like `1031 items in lsp cache`)

**Comment by ofseed** [comment]
I wonder if OOP(specifically, using abstract classes and inheritance) is a viable option. `:h dev-lua` indicates that elaborate OOP design should be avoided.

The benefits of using OOP are as follows:

- Code can be shared through abstract classes and inheritance while keeping flexibility by overriding, or at least, interfaces can be clearly defined. Wherever `bufstate` is used, there is boilerplate code.
- A general benefit of OOP design is to solve the problem of inconsistent state. In our case, I mean all functions that use `bufstate` should be methods of the `bufstate`(or other name, like `FeatureProvider`) object so that we can avoid nil-checks. This can prevent data from being inconsistent with expectations.
- We currently have a fairly successful example in `lsp/semantic_tokens.lua`, which is an OOP implementation managing its own "`bufstate`" (`STHighlighter`).

My rough idea at the moment is to convert these places using `bufstate` into an OOP design, similar to `vim.lsp.semantic_tokens`, and extract reusable parts as an abstract class to reuse code. Moreover, the existence of a unified parent class might share implementation for the following missing feature:

- `partialResultParams`
- `dynamicRegistration`
-  >visibility (so :checkhealth can print something like `1031 items in lsp cache`)

   This can also be done, just like printing active clients by `:checkhealth vim.lsp`(`client.lua` is also based on OOP design).

**Comment by justinmk** [comment]
We can try that. Primary goal is to find some sort of pattern that can be shared. OOP usually isn't necessary since behavior and state can be shared in other ways, but it's not a blocker in this phase.
